File: LY/python/13-17普通数组/15.轮转数组.py
from typing import List
from collections import deque
class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        dq = deque(nums[i] for i in range(len(nums)))
        while k>0:
            cur = dq[-1]
            dq.pop()
            dq.appendleft(cur)
            k-=1
        for i in range(len(dq)):
            nums[i] = dq[i]
class Solution2:
    def rotate(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        for i in range(k):
            cur = nums[-1]
            nums.pop()
            nums.insert(0, cur)
            
sol = Solution2()
nums = [1,2,3,4,5,6,7]
k = 3
sol.rotate(nums, k)
print(nums)

File: WXQ/java/13-17普通数组/15.轮转数组.java
import java.util.*;

import static java.util.Collections.reverse;

public class Test3 {
    public static void main(String[] args) {
        int[] nums={1,2,3,4,5,6,7};
        int k=3;
        Solution solution = new Solution();
        solution.rotate(nums,k);
    }
}
class Solution {
    public void rotate(int[] nums, int k) {
        int n = nums.length;
        k=k%n;
        reverse(nums,0,n-1);//整体翻转
        reverse(nums,0,k-1);//翻转前K个
        reverse(nums,k,n-1);//翻转后K个

    }
    private void reverse(int[] nums, int start, int end) {
        while (start < end) {
            int temp = nums[start];
            nums[start] = nums[end];
            nums[end] = temp;
            start++;
            end--;
        }
    }
}

